[TECH] Normalisation des locales de PixSite sous leur forme canonique (PIX-18715)

lego-technix opened the pull request:
## :pancakes: Problème

Le code de PixSite n'est pas complètement conforme à l'ADR https://github.com/1024pix/pix/blob/dev/docs/adr/0040-locales-languages.md dans le sens où les locales des différents pix-sites ne sont pas définies en forme canonique (c'est à dire `fr-FR`, `fr-BE`, `nl-BE`, etc.). Cela induit notamment des lourdeurs dans le code et des risques car on ne sait jamais vraiment si la locale manipulées en version canonique (comme c'est requis par l'ADR) ou dans une version non-normalisée.

## :bacon: Proposition

La solution proposée est de manipuler dans le code systématiquement des locales sous forme canonique en normalisant la locale uniquement à « l'entrée » à la lecture du cookie de l'utilisateur, car on ne peut avoir confiance dans une valeur en provenance d'un utilisateur.

Ainsi, notamment, on n'est plus obliger de normaliser la valeur de la locale lors de l'écriture dans le cookie.

## 🧃 Remarques

Au niveau de la configuration Nginx, des redirections HTTP permanentes (`fr-be`→`fr-BE`, `nl-be`→`nl-BE`) sont mises en place de manière à rediriger les utilisateurs vers les pages HTML avec des URL en version canonique et que leurs navigateurs gardent cette redirection en mémoire, ce qui évitera aux pix-sites d'exécuter des négociations de locale et des traitements inutiles, économisant ainsi du temps et de l'énergie.

## :yum: Pour tester

:information_source: Tous les tests sont à accomplir en navigation privée pour ne pas être pollué par les données persistantes de son navigateur.

* Vérifier que la présence d'un cookie `locale` redirige bien l'utilisateur sur la version localisée du site, pour cela : 
  1. Se rendre sur Pix Site .org (https://site-pr746.review.pix.org/),
  2. Choisir la locale `Belgique (Français)`,
  3. Constater qu'on arrive bien sur la version correspondant à la locale `fr-BE` du site (c'est à dire sur https://site-pr746.review.pix.org/fr-BE/)
  4. Se rendre à nouveau à la racine de Pix Site .org (https://site-pr746.review.pix.org/), en faisant la modification de l'URL dans la barre d'adresse de son navigateur,
  5. Constater qu'on revient bien sur la version version correspondant à la locale `fr-BE` du site.
* Naviguer parmi les versions en différentes locales et vérifier que les URL sont maintenant toujours en forme canonique, c'est à dire avec des majuscules à la fin pour les formes  suivantes : `fr-FR`, `fr-BE`, `nl-BE`.

Faire enfin une passe de non-régression suivant les différents configurations des Pix Sites : 
* suivant les différents sites : Pix Site et Pix Pro
* suivant les différents domaines : `.fr` et `.org`
* suivant les différents facteurs de taille d'affichage : format grand ou format petit (qui utilise alors le composant _BurgerMenu_)
* avec les différentes locales

En effet les tests de ce projet sont peu couvrants et notamment quasiment absents pour Pix Pro.

lego-technix committed:
feat(shared): use canonical locales by default

lego-technix committed:
feat(shared): use canonical locales by default

lego-technix committed:
feat(pix-site): use canonical locales by default

lego-technix committed:
feat(pix-pro): use canonical locales by default

lego-technix committed:
feat(pix-pro): use canonical locales by default

pix-bot-github commented:
Une fois les applications déployées, elles seront accessibles via les liens suivants :
  * [Pix Site (.fr)](https://site-pr746.review.pix.fr)
  * [Pix Site (.org)](https://site-pr746.review.pix.org)
  * [Pix Pro (.fr)](https://pro-pr746.review.pix.fr)
  * [Pix Pro (.org)](https://pro-pr746.review.pix.org)

Les variables d'environnement seront accessibles sur scalingo https://dashboard.scalingo.com/apps/osc-fr1/pix-site-review-pr746/environment

pix-bot-github commented:
Une fois les applications déployées, elles seront accessibles via les liens suivants :
  * [Pix Site (.fr)](https://site-pr746.review.pix.fr)
  * [Pix Site (.org)](https://site-pr746.review.pix.org)
  * [Pix Pro (.fr)](https://pro-pr746.review.pix.fr)
  * [Pix Pro (.org)](https://pro-pr746.review.pix.org)

Les variables d'environnement seront accessibles sur scalingo https://dashboard.scalingo.com/apps/osc-fr1/pix-site-review-pr746/environment

lego-technix committed:
refactor(shared): simplify now that all handled locales are in canonical form

lego-technix committed:
refactor(shared): simplify now that all handled locales are in canonical form

lego-technix committed:
refactor(pix-site): simplify now that all handled locales are in canonical form

lego-technix committed:
refactor(pix-site): simplify now that all handled locales are in canonical form

lego-technix committed:
refactor(pix-pro): simplify now that all handled locales are in canonical form

lego-technix committed:
refactor(pix-pro): simplify now that all handled locales are in canonical form

lego-technix committed:
refactor(shared): remove now unneeded canonical form normalization

because the given locale comes from the availableLocales which are now all in
canonical form.

lego-technix committed:
refactor(shared): remove now unneeded canonical form normalization

because the given locale comes from the availableLocales which are now all in
canonical form.

pix-bot-github commented:
Une fois les applications déployées, elles seront accessibles via les liens suivants :
  * [Pix Site (.fr)](https://site-pr746.review.pix.fr)
  * [Pix Site (.org)](https://site-pr746.review.pix.org)
  * [Pix Pro (.fr)](https://pro-pr746.review.pix.fr)
  * [Pix Pro (.org)](https://pro-pr746.review.pix.org)

Les variables d'environnement seront accessibles sur scalingo https://dashboard.scalingo.com/apps/osc-fr1/pix-site-review-pr746/environment

theotime2005 commented:
Le titre de la PR ne doit-il pas être en français ?

lego-technix commented:
> Le titre de la PR ne doit-il pas être en français ?

Oui, tu as raison. Titre modifié ainsi : 
> Normalisation des locales de PixSite sous leur forme canonique

lego-technix committed:
feat(pix-site): use canonical locales by default
